clang/lib/Interpreter/IncrementalParser.cpp

```diff
@@ -407,6 +406,16 @@ void IncrementalParser::CleanUpPTU(PartialTranslationUnit &PTU) {
       }
     }
   }
+
+  // FIXME: We should de-allocate MostRecentTU
+  for (Decl *D : MostRecentTU->decls()) {
+    if (!isa<NamedDecl>(D))
+      continue;
+    // Check if we need to clean up the IdResolver chain.
+    NamedDecl *ND = cast<NamedDecl>(D);
```

Comment:
```suggestion
    const auto *ND = dyn_cast<NamedDecl>(D);
    if (!ND)
      continue;
    // Check if we need to clean up the IdResolver chain.
```

Reply:
The `isa` + `cast` saves a few instructions in the case where `D` is not a `NamedDecl`. I'd prefer to stay the way it is although we are not expecting a lot of declarations that are not `NamedDecl`s.

Reply:
isa + cast is a code smell because that is literally doing what dyn_cast does; do you have measurements that suggest there is a measurable performance difference?

Reply:
Hm. Ok I did not realize that code is smelly :)

We are getting a bit theoretical here but for the sake of the argument:

`isa<>` calls [CastInfo<To, const From>::isPossible](https://llvm.org/doxygen/structllvm_1_1CastIsPossible.html#a44e6fa462c45ccec1dcb2e52901e055d)(Val);

`cast<>` calls [CastInfo<To, const From>::doCast](https://llvm.org/doxygen/structllvm_1_1CastInfo.html#a143f6f757917e1245aed3882209b466f)(Val);

`dyn_cast<>` is both `isa + cast`.

My point was that in the case of something that's not a `NamedDecl` we will avoid the check `isPossible` for that case. It's probably peanuts: https://godbolt.org/z/xeh7jh5ax

Anyway, if you feel strongly about the other approach I can do it...

Reply:
If we have performance numbers that suggest this is a hot path and we need the split, then we can keep it. But in general, we [document](https://llvm.org/docs/ProgrammersManual.html#the-isa-cast-and-dyn-cast-templates) that particular pattern:

> "Note that you should not use an isa<> test followed by a cast<>, for that use the dyn_cast<> operator."

Reply:
I don't. Let me make these changes so we can move forward here.

Reply:
OTOH, that loop is easier to be spelled like:
```
for() {
  if (auto *ND = dyn_cast<NamedDecl>(D))
    // do some work
}
```

That violates the loop early exit policy :)